### DISCUSSION TITLE
Keycloak Admin and Account Theme

### DISCUSSION THREAD

**Discussion by lillyqian** [keycloak/keycloak#19508] [Archive]
Our team is starting to use Keyclock as the identity and access management solution for our projects, I found that there are not much flexibility to customize the Admin Console theme as I could with Login. I’d like to make sure that I didn’t miss anything, could anyone confirm that for Admin and Account, I couldn’t customize the theme by template.ftl, for Admin I cannot over write the message by messages_en.properties? 

I didn’t find any helpful information on customizing Admin Console or Account in this document https://www.keycloak.org/docs/latest/server_development/#_themes, but I did see some online posting mentioning that Admin and Account in Keycloak v2 is not customizable anymore as since they were rewritten with React.  It seems  that Account can be customized by overwrite the content of those files  https://github.com/keycloak/keycloak/tree/release/20.0/themes/src/main/resources/theme/keycloak.v2/account, since there is no “admin” directory under https://github.com/keycloak/keycloak/tree/release/20.0/themes/src/main/resources/theme/keycloak.v2, I assume that we couldn’t do much customizing for admin except the css?

**Comment by ssilvert** [comment]
We finally have some time to devote to customization.  We have someone assigned to write documentation on how to do simple theming of admin console.  For more complex changes, it can all be customized, but for now we are advising to fork and modify.

For account console, there is this:
https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension/extend-account-console

However, we will soon be upgrading the account console to a refactored version so the above tutorial probably won't apply.  The goal will be to do something much better than the customization scheme shown in the quickstart.  This will apply to both admin and account consoles.

The great advantage of the Freemarker approach we still use for login is that it is really easy to make changes.  However, the disadvantage is that it's very old technology which can make it very difficult to do more advanced customization.

**Reply by thaitv17** [reply]
Hi @ssilvert , not sure but when i move js folder to another project (in my own project). The admin-ui or keycloak-server did not run anymore. Do you have any ideas?

```
pnpm dev                  


> admin-ui@ dev /home/thaito/Desktop/keycloak-up-extensions/js/apps/admin-ui
> wireit

sh: 1: wireit: not found
 ELIFECYCLE  Command failed.
```

**Comment by lillyqian** [comment]
Thank you for the information @ssilvert. It's great to know that a refactored version of admin/account consoles is coming, do you have a approximate timeline for the release?
Totally agree with you on the Freemarker approach, will the Freemarker approach still be used for admin/account consoles?

**Comment by ssilvert** [comment]
Only account console will see a new refactored version as admin console is already where we want it to be.  Account console V3 should be in the build within a few weeks as a tech preview.  Then we will get rid of account console V1 and V2 as quickly as possible.  

At that point, both account and admin consoles will be using modern, hook-based React.  And both will use Vite instead of Snowpack.  We will document a common approach to customization for both consoles.

Freemarker will not completely go away any time soon, but it will not be used in the consoles except possibly for bootstrapping.

**Reply by lillyqian** [reply]
For security reason, our dev team is not allowed to modify Keycloak source code, is there a way we could customize the messages defined in the json files under ../blob/main/js/apps/admin-ui/public/resources without changing Keycloak source code? 

For example the messages for the Admin dashborad in this file: https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/public/resources/en/dashboard.json

**Reply by jonkoops** [reply]
It might be possible to [create](https://www.keycloak.org/docs/latest/server_development/#creating-a-theme) your own theme and place an asset with the same name in the directory structure. I haven't verified this work though.

**Reply by lillyqian** [reply]
In my project I  put dashboard.json under 
..\themes\ [myCustomeThemeName]\admin\resources\en\  and  under
..\themes\ [myCustomeThemeName]\admin\resources\

Made changes to the dashboard.json in both places, but couldn't see the changes in UI. Do you suggest a different directory?

**Reply by Shuyinsama** [reply]
@ssilvert a follow up question on this since I can't seem to find a clear answer on this. But will there also be the possibility to change "functionality" (custom forms etc)?

For example we want to change the Admin theme specifically in the `Create user` screen where we want custom fields that map directly to attributes. We want our clients to be able to add new users with the necessary requirements without them having to click through stuff like `attributes` so that they cannot make any errors.

We used to have this in the old theme but ever since upgrading to v21 these themes don't work anymore. Downgrading Keycloak is not an option due to CVE's that were fixed. Is it even still possible to use the old themes by setting a flag during startup?

**Reply by ssilvert** [reply]
@Shuyinsama I think it sounds like the user profile feature will do what you want.  The latest version of admin console will automatically generate the custom fields for you on the user details page.  

For account console, the custom attributes are available in keycloak.v3, which is early alpha code right now.  You enable account console v3 with `--features=account3`.

**Comment by ssilvert** [comment]
Did you extend the keycloak.v2 theme in theme.properties?

**Reply by lillyqian** [reply]
Yes, I did. The changes I made in ...\themes\\[myCustomeThemeName]\admin\resources\css\styles.css work well, I also changed the icon.svg and logo.svg successfully.

Now I have issue to make the following changes:
1. **favicon.ico**: I put new favicon.ico under \themes\\[myCustomeThemeName]\admin\resources\img\ but not working. However this solution works for login.
2. **Title**: The default admin title  is "Keycloak Administration Console", is there a way to change it? Login title can be changed in the template.ftl file.
3. **Messages**: Need to figure out a way to change the messages defined in those json files :https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/public/resources/en/dashboard.json

**Reply by guptakaran27** [reply]
hi @lillyqian ,
I m also facing similar issues with the admin theme, but for *1(favicon.ico), keycloak is looking for favicon.svg, instead of favicon.ico
Just change the extension of the file, and it will work. (/themes/myTheme/admin/resources/favicon.svg)

 But I m still unable to change the title of Admin Console, do let me know in case of any updates.

**Reply by lillyqian** [reply]
I put a favicon.svg under .../admin/resources/,  it works like a charm!  thank you @guptakaran27 .
In case you'd like  replace the Keycloak logo in the banner, you can put your logo.svg under .../admin/resources/, and if you'd like to replace the Keycloak logo in the Admin landing page, you can put your icon.svg under .../admin/resources/

**Reply by guptakaran27** [reply]
thanks for the update @lillyqian .
I am unable to change the Title of Admin Tab from "Keycloak Administration UI" to say "Karan's GUI". Overriding the admin-messages.properties file is not working.

**Reply by jonkoops** [reply]
Changing the title of the document is not currently supported, but something we'll be working on in the future.

**Reply by ssilvert** [reply]
> thanks for the update @lillyqian . I am unable to change the Title of Admin Tab from "Keycloak Administration UI" to say "Karan's GUI". Overriding the admin-messages.properties file is not working.

For now, you would need to override the index.ftl file.  This is actually a bug as it should not be hard-coded.  I'm attaching an index.ftl you can use.  It goes in the same directory as your theme.properties file.

[index.zip](https://github.com/keycloak/keycloak/files/11235776/index.zip)

**Reply by lillyqian** [reply]
@ssilvert I put the index.ftl in the desire directory, change the title in the index.ftl. The title got updated in the Admin console, but the page is stacked at "Loading the Admin UI" state.

**Reply by ssilvert** [reply]
Oh, I see the problem.  You will need to pull an `index.ftl` file from the exact version of Keycloak you are using.  So what you need to do is to find `org.keycloak.keycloak-admin-ui-my-version.jar`.   It is located in `lib/lib/main` directory. Use the `index.ftl` file from there and it should work.

That's a total hack though.  We will fix this soon.

**Reply by lillyqian** [reply]
This solution works! Thank you @ssilvert. I need to work in Account Console soon and would rather wait for V3 preview release, is there a release date set for it?

**Reply by ssilvert** [reply]
V3 is now included in the nightly build.  So it will ship with the next regular release.

**Reply by lillyqian** [reply]
That's good news. Our project has a request that not allow user to update their email address in Account Console, do we have the flexibility to remove the email field when user updates his own information in V3?

**Reply by ssilvert** [reply]
You will soon.  The back end is working but hiding a field is not working yet.  So user will see the email address, but update will fail.

**Reply by lillyqian** [reply]
In the Admin Console create user UI, can we make the following changes:
1. Make the email, first name and last name required fields (high priority)
2. Remove/disable Email Verification Settings (medium priority)

In the  user views, can we remove/disable the  Consent and  Attributes Tab? (low priority)

**Comment by ssilvert** [comment]
@lillyqian That should work.  Someone is going to be working on a how-to doc for your use cases and more next week.  If we find use cases like yours that aren't working then we will fix them.

**Reply by lillyqian** [reply]
Could you please post the how-to doc link here when it's available? Thanks!

**Comment by ssilvert** [comment]
We will announce when it's ready.

**Comment by lillyqian** [comment]
The Account Console version 3 is now available as an experimental feature in Keycloak 21.1.0. I'd like to try it out but could find the document on how to costuming the theme, do you have an updated document? @ssilvert @jonkoops @guptakaran27 @abbra

**Reply by jonkoops** [reply]
There is currently no details on how to customize the UI for Account v3. We're planning on making this generic between the Admin and Account UIs so that the guide can apply to both.

**Reply by lillyqian** [reply]
Currently is it just the guide unavailable or the generic solution for Admin and  Account under development? I guess that what I am trying to ask is when I can use those features? :)

**Reply by jonkoops** [reply]
The latter at the moment, we're trying to determine the extend of customization we want to allow and what it's implementation should be.

**Reply by lillyqian** [reply]
In the Admin Console create user UI, we'd like to include those UI changes in our May release, is there a hack for this before the generic solution is release ?

- Make the email, first name and last name required fields (high priority)
- Remove/disable Email Verification Settings (medium priority)

**Reply by jonkoops** [reply]
I think you might be able to use the 'User Profile' feature for the first one (under realm settings), the latter I am not sure what you mean.

**Reply by lillyqian** [reply]
Hi @jonkoops , I'm happy to try the 'User Profile' feature in V3. I upgraded to V21.1.1 and run 
kc.bat start-dev --features=account3, 
but I got the following error, any insight on this?

Without account3 option, it works well:
kc.bat start-dev --features=account3 --spi-theme-static-max-age=-1 --spi-theme-cache-themes=false --spi-theme-cache-templates=false


The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2023-04-24 10:38:02,523 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider

ERROR: Failed to run 'build' command.
ERROR: C:\IoT\SIM\sas-iot-keycloak-theme\customTheme\lib\quarkus\build-system.properties
java.lang.NullPointerException
        at io.smallrye.config.ExpressionConfigSourceInterceptor.escapeDollarIfExists(ExpressionConfigSourceInterceptor.java:89)
        at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:61)
        at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:42)
        at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
        at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:17)
        at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
        at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:324)
        at org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler.printErrorHints(ExecutionExceptionHandler.java:102)
        at org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler.dumpException(ExecutionExceptionHandler.java:92)
        at org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler.error(ExecutionExceptionHandler.java:75)
        at org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler.handleExecutionException(ExecutionExceptionHandler.java:45)
        at picocli.CommandLine.execute(CommandLine.java:2088)
        at org.keycloak.quarkus.runtime.cli.Picocli.runReAugmentation(Picocli.java:184)
        at org.keycloak.quarkus.runtime.cli.Picocli.runReAugmentationIfNeeded(Picocli.java:121)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:89)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)

**Reply by lillyqian** [reply]
Regarding to "Remove/disable Email Verification Settings": when creating a new user, there is an "Email verified" option, is there a way to removing it or disable it from the UI?

**Reply by jonkoops** [reply]
Hmmm, I am not familiar with this exception. Perhaps this is because the UI was excluded from the Quarkus dist? I remember a bug for it beign closed (should be fixed in the next patch release).

As far as I know there is no way to remove the "Email verified" option.

**Reply by lillyqian** [reply]
I upgraded to the new release 21.1.1 and the above error goes away. Under "Realm Settings", I don't see the User Profile' feature, is there any feature I need to turn on additionally?

![image](https://user-images.githubusercontent.com/26976350/234690277-5b8bdd47-5b3e-4f17-b4a0-2bb326048082.png)

![image](https://user-images.githubusercontent.com/26976350/234690106-f16bfc2e-4cef-4e1b-97fb-153c1b794c87.png)

**Reply by jonkoops** [reply]
@lillyqian Yes, this feature is still behind a profile. If you start the server with `--features=declarative-user-profile` you will see a toggle under 'Realm settings' :arrow_right: 'General':

!["User Profile Enabled" switch](https://user-images.githubusercontent.com/695720/235638398-da7d552c-a6a0-4cd9-bc7c-276147695f27.png)

**Reply by lillyqian** [reply]
The user profile feature looks great! I understand this is a experimental feature for now, do you recommend to use it in the production environment?

The username in the create new user window is a required field but it is not marked with an asterisk---maybe a small bug?

**Reply by ssilvert** [reply]
I think it is very unlikely that user profile will go away.  It's actually in `Preview` status right now, which is a step above `Experimental`. It's just my opinion, but I wouldn't be afraid to use it in production.

Maybe a small bug.  When you are using email as username, then username is not required.  Username field really needs to be hidden in that case.  Do you mind filing an issue for it?

**Reply by lillyqian** [reply]
I'd happy to fill an issue. do you have a link? 
What is the lowest version that supports this user profile feature?

**Reply by ssilvert** [reply]
https://github.com/keycloak/keycloak/issues

I don't know the lowest version.  To find out, you could probably check the Keycloak blog or look through release notes.

**Reply by krasholo** [reply]
hi all!!!

@ssilvert news about the how-to documentations to customize the new admin-ui console?

**Reply by ssilvert** [reply]
First commit just happened.  Much more to come.
https://github.com/keycloak/keycloak-quickstarts/tree/latest/extend-admin-console-quick-start

**Reply by lillyqian** [reply]
@ssilvert Is there a way to customize the messages like can do with the Login using ...\login\messages\messages_en.properties file?

**Reply by lillyqian** [reply]
> @lillyqian Yes, this feature is still behind a profile. If you start the server with `--features=declarative-user-profile` you will see a toggle under 'Realm settings' ➡️ 'General':
> 
> !["User Profile Enabled" switch](https://user-images.githubusercontent.com/695720/235638398-da7d552c-a6a0-4cd9-bc7c-276147695f27.png)

@jonkoops @ssilvert Is there a date set for including this user profile preview feature in a formal release? We have to use formal release feature in our production for security reason. We would like to know the time if possible so we can plan our work. Thank you.

**Reply by ssilvert** [reply]
@lillyqian The User Profile feature is there in all recent releases.  It's just in preview mode so you have to enable it with a feature flag.

**Reply by lillyqian** [reply]
I can enable it in our dev environment,   but unfortunately preview mode features are not allowed to be used in our production for security reason.  when will it be changed from a preview mode to a formal release mode?

**Reply by ssilvert** [reply]
It's hard to say when it will go out of preview.  I just can't give you a good answer here.

**Reply by initshdb** [reply]
@ssilvert We have a similar requirement where we want to customize the admin-ui. 

**Requirement**
- Certain users are given hardware HMAC tokens for Two Factor, it should be possible for them to login using the OTP from the tokens. We have a customized provider and authenticator around this. The HMAC secret and other HMAC configuration details can only be set by the admin user. 
- With earlier KC old UI we customized the underlying Angular JS files (html, services, and controllers) relevant to the user-credentials tab where we add our own form for the admin to enter and save the HMAC details (secret, no. of OTP digits, etc.) for a user.
- How can we achieve the same use case with the latest version based on React. Will the approach in **keycloak-quickstarts** for [extend-account-console](https://github.com/keycloak/keycloak-quickstarts/tree/latest/extend-account-console) work where there is a new tab added with navigation to new React JSX/JS pages. Or is there another way possible? 
- Also is it advisable to override/customize the React files in our custom theme folder. And if so should we use the folder theme approach `KC/themes/custom-theme/admin/` or create a jar of custom theme (`custom-admin-ui.jar`)

Please guide and suggest.

**Reply by jonkoops** [reply]
> work where there is a new tab added with navigation to new React JSX/JS pages

This is what we are looking to accomplish, or at least something in that direction. However we currently do not have a technical solution in place.

**Comment by ckrc** [comment]


**Reply by ckrc** [reply]
Problem resolved - I found out how to build the jar file properly (Install maven and refer to the README.md file under ./keycloak-theme folder), and replacing the file in `/opt/keycloak/lib/lib/main` is correct (making sure the file name remains the same though)

**Comment by kusuma-ms** [comment]
HI, 
Not able to access any of the links specified above it says not found . Is the repo link changed ?

**Reply by ssilvert** [reply]
Exactly which links are not working?

**Reply by DaBonse** [reply]
I think, kusuma-ms means this link:

https://github.com/keycloak/keycloak-quickstarts/tree/latest/extend-account-console

Version latest and 22 does not exist, last version ot the tutorial is 21.1.

**Reply by ssilvert** [reply]
Ah, yes.  The location has changed since I wrote that.  I update the comment above.  The new location is https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension/extend-account-console

**Reply by AlexeyZhevnyak** [reply]
Hi! Is there a way to extend admin console with adding some new functionality, not only changing css and icons?

**Reply by ssilvert** [reply]
Right now, the only solution is to fork and modify the whole application.  But we are in the process of creating a better way to do it.

**Reply by AlexeyZhevnyak** [reply]
Will it work, if i build keycloak-admin-ui.jar and add it to the /providers folder, as it says in the documentation?
![image](https://github.com/keycloak/keycloak/assets/53194024/2033fd7e-7ad9-4077-b96c-33f913b634d9)

**Reply by jonkoops** [reply]
Yes, that should work indeed.

**Reply by daniel-pimenta-DME** [reply]
@AlexeyZhevnyak 
I'm looking to change some javascript logic on a page of Admin UI
Did you try to add that jar on providers folder? Did it work?
I've done it for a custom theme but wondering if it can be done the same way for changing ReactJS code.

**Reply by AlexeyZhevnyak** [reply]
i did this, but i don't see any changes in my admin ui

**Reply by daniel-pimenta-DME** [reply]
it did work for me
i built keycloak from source, then copied admin ui from js/apps/admin-ui/target/keycloak-admin-ui-version.jar to /opt/keycloak/providers and it loaded the changed admin-ui.

**Comment by kusuma-ms** [comment]
Hi All,

Thanks for the information and suggestion.
I am not able to add new theme for Admin page, i I add the , the keycloak application crashes . Even if I select the existing theme i.e. basic for Admin page it crashes.

How to overcome this issue ?

**Comment by a9ki21023** [comment]


**Reply by jonkoops** [reply]
This seems like an unrelated discussion to the original post, I would recommend that you start a new discussion instead.

**Comment by ThinkFactoryDigital** [comment]
I tried to use the logoUrl of the client in the account console, but can not get it. As far as I see the content therefore comes from application.tsx and there is no one of this (tosUri, policyUri and logoUri included?

 {
"clientId": "test-client",
"clientName": "Test Application",
"description": "This is a test application",
"userConsentRequired": false,
"inUse": false,
"offlineAccess": false,
"rootUrl": "http://localhost/testclient",
"baseUrl": "",
"effectiveUrl": ""
},
{
"clientId": "account-console",
"clientName": "Account Console",
"userConsentRequired": false,
"inUse": true,
"offlineAccess": false,
"rootUrl": "${authBaseUrl}",
"baseUrl": "/realms/myrealm/account/",
"effectiveUrl": "http://localhost:8080/auth/realms/myrealm/account/"
},

Happend this because it was in keycloak.v2.account ... I tried to change the theme, but I do not get a logoUri also with the account3 feature??? Am I wrong or did n't I see something important? 

Also I found this: 
![image](https://github.com/keycloak/keycloak/assets/3061619/5af12a92-8ef7-40a7-bb21-bb0fb83b6556)

Is it possible that old changes dissapeared again, because[ this topic](https://github.com/keycloak/keycloak/pull/11837) was closed in 2022.

**Reply by cgeorgilakis** [reply]
Do these clients have consent? Have these value configured?

These values will be shown in consent and will be shown in account console only if User has already accepted a consent.
As I see from [AccountRestService class](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java) line 307-309 code still exists.

**Comment by ThinkFactoryDigital** [comment]
You mean I have to activate this 
![image](https://github.com/keycloak/keycloak/assets/3061619/8e19ea26-e334-4aa3-b076-8bdfbd63470b)
for every application I want to show the client logo and each time a user has to accept this? I tried it and I got a request, if I will accept this, but nevertheless no logoUri inside the response 
![image](https://github.com/keycloak/keycloak/assets/3061619/1f2f3168-1abd-4977-8cb5-48d7d7219c14)

**Reply by jonkoops** [reply]
@crowmedia please reply in a thread rather than starting a new one.

**Comment by ThinkFactoryDigital** [comment]
@cgeorgilakis I think I found out what you mean, but that was not the result I expected.  I did not want a consent of the user for a client but I wanted to add a client logo to the list of my applications in the account console. For example here 
![image](https://github.com/keycloak/keycloak/assets/3061619/9f38d269-86a4-4483-a5de-2607a5160729)

**Reply by jonkoops** [reply]
@crowmedia please reply in a thread rather than starting a new one.

**Comment by ThinkFactoryDigital** [comment]
@cgeorgilakis : After I added this consent for example to 3 clients I got just a request for the application console, and I also get there now the logoUri, but just for appliction console.. To be honest I do not really understand why I need a consent for a logoUri. TermsOfUse for example I can understand. I wanted to use the endpoint to extend keycloak and our theme, but I think I have to write my own endpoint now.   Or do you see another solution?

**Reply by jonkoops** [reply]
@crowmedia please reply in a thread rather than starting a new one.

**Comment by jerrycooper** [comment]
Hi,
Is a way to add custom message to admin theme?
Because during addition password policy in admin-ui, there is help icon with text. 
I didn't find a way to add help text in UI to our policy. 
Original text are here [https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages](https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages)

**Reply by danielFesenmeyer** [reply]
You could either inherit admin theme with your own resources (properties), like shown here. https://www.keycloak.org/docs/latest/server_development/index.html#_theme_resource

If you just want to overwrite the text, it's probably easier to use "realm localization texts", as (briefly) mentioned here: https://www.keycloak.org/docs/latest/server_admin/#:~:text=All%20internationalized%20texts%20which%20are%20provided%20by%20the%20theme%20can%20be%20overwritten%20by%20realm%2Dspecific%20texts%20on%20the%20Localization%20tab

**Reply by jerrycooper** [reply]
I have my custom theme, but message is not translated. It looks like the text goes from locale text files in keycloak-admin-ui.jar.
Policies texts are in authentication-help.json.
Is it possible overwrite texts in keycloak-admin-ui.jar?

**Reply by fari-99** [reply]
if my understanding is correct, you can override the text using [messages](https://www.keycloak.org/docs/latest/server_development/index.html#messages).

**Comment by LukeIntegral** [comment]
Hi,

We have noticed that the Admin Console and Themes v1 have been removed from the build. V3 has been delivered but hardly any documentation on how to customise it exists, is there a plan to improve/supply this documentation?

We have noticed that you have removed all documentation and downloads for versions before v21. I am concerned it doesn’t appear we even have access to our current version. Is there a reason why you have done this?

**Reply by fari-99** [reply]
it seems there is a team that will create a customization easier for admin and account console [here](https://github.com/keycloak/keycloak/discussions/19508#discussioncomment-6929459). i have the same problem, waiting for the new documentation and feature for this customization easier.

for now i will try to use [this](https://github.com/keycloak/keycloak-quickstarts/tree/23.0.1/extension/extend-account-console) for account console, and [this](https://github.com/keycloak/keycloak-quickstarts/tree/23.0.1/extension/extend-admin-console) for admin console. don't know if can be used on latest version tho.

**Comment by kouta-kiriyama** [comment]
Hi,

I  make custom theme. 
`themes/src/main/resources/theme/mytheme`

and this statement
https://github.com/keycloak/keycloak-quickstarts/blob/latest/extension/extend-admin-console/README.md
`Keycloak offers the --spi-theme-folder-dir=/path/to/themes flag which enables the specification of the directory where the custom theme is located, allowing for changes to be made directly to the source location without the need for copying files to the server. This can used by starting Keycloak server with the --spi-theme-folder-dir=/path/to/themes flag, and replacing /path/to/themes with the directory path of the custom theme.`

so I trid 
'kc.bat start-dev --spi-theme-folder-dir=\~/themes'
but mytheme can't be selected.
'kc.bat start-dev --spi-theme-folder-dir=\~/themes/src/main/resources/theme'
custome theme can be selected. But Localization (Supported locales) can't be selected (only English).
![image](https://github.com/keycloak/keycloak/assets/107161280/80073004-f00f-4065-a1ab-6c19046bead3)

I want to select locale and change mytheme source ........

**Comment by mkopeyka** [comment]
I have to migrate from version `20.0.2` and I'm stuck with admin theme customization.
I have disabled some controls in old admin theme to avoid adding, editing or deleting of user attribute.
How to implement this restriction in keycloak `25.0.1`?

**Comment by ajaiswal595** [comment]
@ssilvert Can you please advise here

[https://keycloak.discourse.group/t/account-console-theme-migration-from-24-0-4-to-25-0-6](https://keycloak.discourse.group/t/account-console-theme-migration-from-24-0-4-to-25-0-6/28195)

**Reply by ssilvert** [reply]
@ajaiswal595 The account console is now under /js/apps/account-ui

**Reply by ajaiswal595** [reply]
Thanks @ssilvert 
The answer was late, I have already migrated the account theme 

But there are a couple of issues I have observed 

1.Why did we  change favicon.ico to favicon.svg 
2. What is the use of the Jump to section it doesn't do anything.
<img width="1183" alt="image" src="https://github.com/user-attachments/assets/3a0c00e3-f1f0-4070-b3cb-c8e0a868379b">
3. Clients are broken or expected it shows like this ${client_account-console}, ${client_security-admin-console}
<img width="1356" alt="image" src="https://github.com/user-attachments/assets/d4759833-a070-4ca2-960b-67eb469fa698">

**Comment by ssilvert** [comment]
@ajaiswal595 
For #1, we changed to svg because it's a better format.  You can Google to find out why.
For #2, please create an issue for this.  It's a bug. If there is only one section left, you are correct that the jumps don't make sense.
For #3, I think this was fixed in Keycloak 26.  If not, please create an issue for it.
